UnitTests/View/Adornment/AdornmentTests.cs

```diff
@@ -15,35 +15,80 @@ public void Bounds_Location_Always_Empty_Size_Correct ()
             X = 1,
             Y = 2,
             Width = 20,
-            Height = 31
+            Height = 20
         };
 
+        view.BeginInit ();
+        view.EndInit ();
+
+        Assert.Equal (new (1, 2, 20, 20), view.Frame);
+        Assert.Equal (new (0, 0, 20, 20), view.Bounds);
+
         var marginThickness = 1;
-        view.Margin.Thickness = new Thickness (marginThickness);
+        view.Margin.Thickness = new  (marginThickness);
+        Assert.Equal (new (0, 0, 18, 18), view.Bounds);
 
         var borderThickness = 2;
-        view.Border.Thickness = new Thickness (borderThickness);
+        view.Border.Thickness = new (borderThickness);
+        Assert.Equal (new (0, 0, 14, 14), view.Bounds);
 
         var paddingThickness = 3;
         view.Padding.Thickness = new Thickness (paddingThickness);
+        Assert.Equal (new (0, 0, 8, 8), view.Bounds);
 
-        view.BeginInit ();
-        view.EndInit ();
+        Assert.Equal (new (0, 0, view.Margin.Frame.Width, view.Margin.Frame.Height), view.Margin.Bounds);
 
-        Assert.Equal (new Rectangle (1, 2, 20, 31), view.Frame);
-        Assert.Equal (new Rectangle (0, 0, 8, 19), view.Bounds);
+        Assert.Equal (new (0, 0, view.Border.Frame.Width, view.Border.Frame.Height), view.Border.Bounds);
 
-        Assert.Equal (new Rectangle (0, 0, view.Margin.Frame.Width - marginThickness * 2, view.Margin.Frame.Height - marginThickness * 2), view.Margin.Bounds);
+        Assert.Equal (new (0, 0, view.Padding.Frame.Width , view.Padding.Frame.Height), view.Padding.Bounds);
+    }
 
-        Assert.Equal (new Rectangle (0, 0, view.Border.Frame.Width - borderThickness * 2, view.Border.Frame.Height - borderThickness * 2), view.Border.Bounds);
+    // Test that Adornment.Bounds_get override returns Frame.Size minus Thickness
+    [Theory]
+    [InlineData (0, 0, 0, 0, 0)]
```

Comment:
Excellent place for use of the generators in the combinatorial extension or a generator method if that's too annoying to adapt.

Generator method allows shorter test code, too, since you can assert on the rectangles and points, rather than each individual property.

Reply:
This particular one is already a pretty well-written test, so the CombinatorialData thing is a nice fit here.

Reply:
I just haven't had the time to get my head around those things yet.

Reply:
For a lot of our existing tests, it would require some refactoring of the tests themselves.

This one, however, does not take specific "expected" values per case, instead using the inputs to create the expected values (which is the ideal way).

To use that stuff, the method itself needs the `[CombinatorialData]` attribute, and then each parameter gets the appropriate attribute to supply it either with explicit values, a range, or a generator explicitly for just that parameter.[^1]

It then creates the full set of possible cases represented by the cross-join of all of them.

I've done a few in some of the stuff I did for Color, if you want some actual examples.

Just remember it's combinatorial by default. So a method with 5 parameters and 2 values for each means 2^5 = 32 cases. That same test method with ranges covering 2, 3, 4, 5, and 6 values for each, respectively, means 2 * 3 * 4 * 5 * 6 = 720 cases. Or ranges from -10 to 10 for each means 21^5 = 4,084,101 cases... You get the idea.... And Visual studio does NOT behave well when there are that many test cases, and it is also pretty dumb about how it caches all of that, too. 

[^1]: NUnit has all this natively and it looks like the creator of this extension did try to at least mostly emulate the syntax and behavior, so that's cool.